### ISSUE TITLE
How to calculate the FPS of the YOLOv8s model?

### ISSUE THREAD

**Issue by remeberWei** [ultralytics/ultralytics#2690]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Guys, how to calculate the FPS of YOLOv8s model? I run with this command line format (yolo val data=VOC.yaml split=test), and the FPS I got is very poor, I think there is something wrong，My graphics card is 3090. The results obtained by the initial model are shown in the figure below. I think there is a problem.
![image](https://github.com/ultralytics/ultralytics/assets/91881703/7a48c966-714a-4759-a8a3-b1ab09a05ca9)

**Comment by glenn-jocher** [comment]
@remeberWei hi there, to calculate the FPS of the YOLOv8 model, you can use the `--task test --weights <YOUR WEIGHTS FILE> --img 640` command. This command runs the model on the validation dataset with an image size of 640 pixels. The FPS can be calculated as the number of images processed per second. It's worth noting that the performance of the model can be affected by various factors, such as the GPU used, the batch size, and the image size. So, tweaking these hyperparameters may improve the FPS. Let me know if that helps!

**Comment by remeberWei** [comment]
ok，thanks guys.



---Original---
From: "Glenn ***@***.***&gt;
Date: Fri, May 19, 2023 09:59 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] How to calculate the FPS of the YOLOv8smodel? (Issue #2690)




 
@remeberWei hi there, to calculate the FPS of the YOLOv8 model, you can use the --task test --weights <YOUR WEIGHTS FILE&gt; --img 640 command. This command runs the model on the validation dataset with an image size of 640 pixels. The FPS can be calculated as the number of images processed per second. It's worth noting that the performance of the model can be affected by various factors, such as the GPU used, the batch size, and the image size. So, tweaking these hyperparameters may improve the FPS. Let me know if that helps!

**Comment by glenn-jocher** [comment]
Hello @remeberWei, to calculate the FPS of YOLOv8 model, you can use the following command: `--task test --weights <YOUR WEIGHTS FILE> --img 640`. This command runs the model on the validation dataset with an image size of 640 pixels. To calculate the FPS, you can divide the number of images processed by the time taken to process them. It's important to note that the performance of the model is influenced by various factors, such as GPU, batch size, image size, among others. Tweaking these hyperparameters may improve the model's FPS. Let me know if you have any additional questions or concerns.

**Comment by swe-cognitica** [comment]
How to caculate on live video while using cv2 to do some operations?

**Comment by shengmouget** [comment]
The FPS value running on my 12G 4070TI host is like below, is this normal? Or does it mean that there is something wrong with my calculations?
<img width="680" alt="截屏2023-08-04 20 35 36" src="https://github.com/ultralytics/ultralytics/assets/61001480/37af011a-9f07-4ef7-ac26-55e694e74032">

**Comment by glenn-jocher** [comment]
@shengmouget hello, the FPS value is dependent on a variety of factors such as the complexity of your model, the size of your input images, your GPU capacity, and the batch size you are using. If you're running YOLOv8 on a NVIDIA 4070 Ti, keep in mind the utilization of the GPU can significantly impact the FPS you observe.

In case of lower FPS, it doesn't necessarily indicate there's something wrong with your implementation or calculations. You might be running the model with large image sizes or large batch sizes, which can slow down the processing speed. 

Lastly, maintaining sufficient GPU and CPU utilization during inference time is key for higher FPS. Make sure that your GPU is being fully utilized when running the model.

Remember that this is a general guideline and your specific setup might require additional tuning or adjustments. If you believe there could be a problem with your setup, you might consider consulting relevant resources or the community discussions for further assistance.

**Comment by Zhangj11** [comment]
Can you give a concrete code example of yolov8 testing fps? I see that there is no option for task=test in model.val().

**Comment by glenn-jocher** [comment]
@Zhangj11 hello, thank you for reaching out and asking for clarity.

Currently, the `model.val()` function is mainly for evaluating the model performance on a validation dataset in terms of metrics such as mAP (mean Average Precision) rather than for testing the FPS.

To compute the FPS, you'd need to implement a separate code where typically you would keep track of the time before and after the model's inference on a batch of images or a video sequence. The FPS can then be calculated by dividing the total number of batches (or frames if it's a video) by the total time taken.

Please note that several factors may affect the FPS calculation, including the batch size, image size, the model's complexity, and the computational resources available for the task. Also, remember that different settings can lead to different FPS, so make sure to configure your test according to your specific needs.

I hope this explanation provides some directions on how to proceed with your FPS calculations. If you have any more questions, please feel free to ask.

**Comment by abcde-bit** [comment]
@shengmouget hello ,brother.where did you get this picture? Is this what comes with the official YOLO code?

**Comment by abcde-bit** [comment]
@glenn-jocher Hi,I don't understand how to use the folowing "--task test --weights <YOUR WEIGHTS FILE> --img 640". Can you replenish completely？Dose it like yolo test weights=weightfle img=640. I would appreciate it if you could answer this question

**Comment by shengmouget** [comment]
> @shengmouget hello ,brother.where did you get this picture? Is this what comes with the official YOLO code?@shengmouget 你好，兄弟。你从哪里得到这张照片的？这是YOLO官方代码附带的吗？
这是我自己改进的。

**Comment by glenn-jocher** [comment]
@shengmouget to run a test using the command line with YOLOv8 and your specified weights and image size, you would use the CLI command structured as follows:

```yolo val --weights <YOUR WEIGHTS FILE> --img 640 --data <YOUR DATA YAML FILE> --task test```

Please replace `<YOUR WEIGHTS FILE>` with the actual path to your model weights (e.g., `yolov8s.pt`) and `<YOUR DATA YAML FILE>` with the path to your data configuration file (e.g., `VOC.yaml`) as per your dataset.

This command will execute the test task, using the model weights you specified, and evaluate it on the dataset configured in your YAML file with the given image size (`--img 640`).

If there is any confusion or further questions, feel free to post on GitHub discussions or issues for more detailed support.

**Comment by abcde-bit** [comment]
@abcde-bit 能分享一下代码吗，大佬，如果不方便的话可以加我QQ2312428043

**Comment by glenn-jocher** [comment]
@abcde-bit 您好！很抱歉，这里主要侧重于提供关于YOLOv8的官方支持和讨论。关于代码分享和个人通讯的请求，建议在官方渠道外进行，例如通过GitHub分享您的改动，请确保遵守相关许可协议。非常感谢您的理解和支持！🙏

**Comment by Johnny-zbb** [comment]
@abcde-bit Hello! 
I tried：`! yolo detect test model=runs/detect/train/weights/best.pt `，But I get the following tips ：SyntaxError:` 'test' is not a valid YOLO argument. Similar arguments are i.e. ['detect'].
I look forward to your reply.

**Comment by DecipherData** [comment]
To calculate FPS, should we take the number of batches// inference time?
For example,
I have the following output:
val: Scanning /content/drive/data/val/labels.cache... 1175 images, 0 backgrounds, 0 corrupt: 100%|██████████| 1175/1175 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 74/74 [00:44<00:00,  1.65it/s]

                   all       1175       2392      0.992      0.753      0.863      0.862

Speed: 0.3ms preprocess, 4.6ms inference, 0.0ms loss, 1.4ms postprocess per image

Here the number of batches looks like 74. So, should I take just inference time to calculate FPS= 74/4.6 = 16.09 ? or should I add all the time values 0.03+4.6+0.0+1.4=6.03 that makes FPS= 74/6.03=12.27

Or is there any other way? Please guide.

**Comment by glenn-jocher** [comment]
@DecipherData hello! 😊

To calculate FPS (Frames Per Second), you should focus on the inference time per image since it represents the model's ability to process images within a second. In your case, the inference time per image is 4.6ms. To convert this into FPS, you can use the formula:

```python
FPS = 1 / (inference_time_in_seconds_per_image)
```

Given your inference time is 4.6ms (which is 0.0046 seconds), the calculation would be:

```python
FPS = 1 / 0.0046 = ~217
```

This means your model can process approximately 217 frames per second based on the inference time alone.

The other times (preprocess, loss, postprocess) could be considered for a more comprehensive "end-to-end" FPS calculation but typically, when discussing model performance, we focus on the inference time.

Hope this helps! Let me know if you have any more questions.

**Comment by marouen88** [comment]
@glenn-jocher Hello Glenn, I would like to ask you how I can calculate the latency time, or, as already @DecipherData  mentioned, you have to add the time.
Thank you for your help!

**Comment by glenn-jocher** [comment]
To calculate the latency time, you should consider the total time taken for preprocessing, inference, and postprocessing. In your case, adding these times will give you the overall latency per image. If you need further assistance, please ensure you are using the latest package versions to check for any improvements or changes. Thank you!

**Comment by marouen88** [comment]
@glenn-jocher thank you for your help, I would like to ask you what does the class background mean although i have only trained the class person 
![confusion_matrix (1)](https://github.com/user-attachments/assets/dca553a8-0102-4b21-8e1a-4cc1fcb2ff35)

**Comment by glenn-jocher** [comment]
@marouen88 the "background" class typically represents areas in the image that do not belong to any of the defined object classes. In your case, it seems to be a default class for non-person regions. The number 25 likely represents the true positive detections for the "person" class. If you have further questions, please ensure you are using the latest package version to check for any improvements or changes. Thank you!

**Comment by marouen88** [comment]
@glenn-jocher True Positive oder True Negative!!! Thank you!

**Comment by glenn-jocher** [comment]
@marouen88 512 is the person TPs, 25 is the FPs, the predictions that it thought were a person but that were actually background.

**Comment by marouen88** [comment]
@glenn-jocher I would like to ask you whether my model contains an overfitting? Thank you!
![results (1)](https://github.com/user-attachments/assets/a52faac8-0c58-4902-a872-94cec78eff84)

**Comment by glenn-jocher** [comment]
It's difficult to determine overfitting from the provided image alone. Please compare your training and validation loss curves; if the validation loss increases while the training loss decreases, it may indicate overfitting.
